lib/schemas/src/prompts.ts

```diff
@@ -19,11 +19,19 @@ export type SchemaPromptGenerator<T> = (
   // bizOpsSystem will be undefined if project is not a system
   bizOpsSystem?: BizOpsSystem
 ) => Promise<T | undefined>
+
+// HACK:KB:20241119 we use .refine to warn about moved plugin options but each call
+// to .refine wraps the Zod schema type in a ZodEffects which breaks the infer below
+type NestedZodEffects<T extends z.ZodTypeAny> =
```

Comment:
**suggestion**: we can use a recursive `infer` here to allow for an arbitrary number of `.refine` calls. we can also bundle the `infer` for the inner object shape here which allows us to simplify the `PromptGenerators` definition because TypeScript seems to infer the type constraints better!

```ts
type InferZodShape<T> = T extends z.ZodObject<infer Shape>
  ? Shape
  : // we use .refine to warn about moved plugin options but each call to .refine
  // wraps the Zod schema type in a ZodEffects which we need to unwrap first
  T extends z.ZodEffects<infer Unwrapped>
  ? InferZodShape<Unwrapped>
  : never

// This type defines an interface you can use to export prompt generators. The
// `T` type parameter should be the type of your `Schema` object, and it will
// be mapped into a partial object of `SchemaPromptGenerator` functions with
// all their return types set to the output type of each option schema.
export type PromptGenerators<T> = {
  [option in keyof InferZodShape<T>]?: SchemaPromptGenerator<z.output<InferZodShape<T>[option]>>
}
```

Reply:
amended the original commit with this, thanks!